deps.go

```diff
@@ -0,0 +1,236 @@
+		// can be done at a higher level.
+		return nil, nil, nil
+	}
+	args := []string{"list", "-deps", "-json"}
+	if withTests {
+		args = append(args, "-test")
+	}
+	args = append(args, pkgNames...)
+	cmd := exec.Command("go", args...)
+	cmd.Stderr = os.Stderr
+	var buf bytes.Buffer
+	cmd.Stdout = &buf
+	if err := cmd.Run(); err != nil {
+		return nil, nil, errors.Wrap(err)
+	}
+	var pkgs []*Package
+	dec := json.NewDecoder(&buf)
+	for {
+		var pkg Package
+		if err := dec.Decode(&pkg); err != nil {
+			if err != io.EOF {
+				return nil, nil, errors.Wrap(err)
+			}
+			break
+		}
+		pkgs = append(pkgs, &pkg)
+	}
+	pkgMap := make(map[string]*Package)
+	for _, p := range pkgs {
+		pkgMap[p.ImportPath] = p
+	}
+	// deps holds a map from package import path to whether that package
+	// is used for testing only (true) or just the build (false)
+	deps := make(map[string]bool)
+	for _, p := range pkgs {
+		if !p.DepOnly {
+			if err := addDeps(p, pkgMap, isTestPackage(p), true, withTests, deps); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+	for ipName, isTest := range deps {
+		p := pkgMap[ipName]
+		if !p.DepOnly {
+			// It's one of the initial packages, so we don't count it
+			// as a dependency.
+			continue
+		}
+		if isInternal(p.ImportPath) {
+			// It's an internal package - ignore it.
+			continue
+		}
+		if isTest {
+			test = append(test, p)
+		} else {
+			build = append(build, p)
+		}
+	}
+	sort.Slice(build, func(i, j int) bool {
+		return build[i].ImportPath < build[j].ImportPath
+	})
+	sort.Slice(test, func(i, j int) bool {
+		return test[i].ImportPath < test[j].ImportPath
+	})
+	return build, test, nil
+}
+
+func isTestPackage(p *Package) bool {
+	if p.ForTest != "" {
+		return true
+	}
+	if !strings.HasSuffix(p.ImportPath, ".test") {
+		return false
+	}
+	// The pseudo-package created by the go tool
+	// has a Go file that looks like an absolute path and
+	// doesn't have a go suffix.
+	// TODO is there a better way to check this?
+	for _, f := range p.GoFiles {
+		if filepath.IsAbs(f) || !strings.HasSuffix(f, ".go") {
+			return true
+		}
+	}
+	return false
+}
+
+func isInternal(importPath string) bool {
+	// TODO more efficient check.
+	return strings.HasPrefix(importPath, "internal/") || strings.Contains(importPath, "/internal/") || strings.HasSuffix(importPath, "/internal")
+}
+
+func addDeps(p *Package, pkgs map[string]*Package, isTest, isRoot bool, withTests bool, deps map[string]bool) error {
+	if isTest {
+		if _, ok := deps[p.ImportPath]; ok {
+			// It's already present as a build or test dependency.
+			return nil
+		}
+		deps[p.ImportPath] = true
+	} else {
+		if isTest, ok := deps[p.ImportPath]; ok && !isTest {
+			// It's already present as a build dependency.
+			return nil
+		}
+		deps[p.ImportPath] = false
+	}
+	addImports := func(ipNames []string, isTest bool) error {
+		for _, ipName := range ipNames {
+			if ipName == "unsafe" && p.Standard {
+				// Stdlib is always allowed access to unsafe.
```

Comment:
nit: I'd either say `std` or `Standard library`, since those are AFAIK the only official names

Reply:
done